### PR TITLE
[ASoC 2022] Pixiu Metrics Implementation #433

### PR DESCRIPTION
**Pull request by gongna-au**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #433

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pixiu/pkg/filter/prometheus/metric._test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+	"bytes"
+	"encoding/json"
+	"net/http"
+	"testing"
+)
+import (
+	"github.com/apache/dubbo-go-pixiu/pixiu/pkg/common/extension/filter"
+	"github.com/apache/dubbo-go-pixiu/pixiu/pkg/common/yaml"
+	"github.com/apache/dubbo-go-pixiu/pixiu/pkg/context/mock"
+	"github.com/stretchr/testify/assert"
```
Comment (baerwang) [inline pixiu/pkg/filter/prometheus/metric._test.go]: import style

File: pixiu/pkg/filter/prometheus/config.go [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package prometheus
+
+type (
+	MetricCollectConfiguration struct {
+		Rules MetricCollectRule `yaml:"metric_collect_rules" json:"metric_collect_rules"`
+	}
+	MetricCollectRule struct {
```
Comment (baerwang) [inline pixiu/pkg/filter/prometheus/config.go:25]: one grid apart

**Review by AlexStocks — APPROVED** [review]

File: pixiu/pkg/filter/prometheus/metric.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	if f.Prom == nil {
+		logger.Errorf("Message:Prometheus Collector is not initialized")
+		ctx.SendLocalReply(nh.StatusForbidden, constant.Default403Body)
+		return filter.Continue
+	}
+
+	f.Prom.SetMetricPath(f.Cfg.Rules.MeticPath)
+	f.Prom.SetPushGateway(f.Cfg.Rules.PushGatewayURL, time.Duration(f.Cfg.Rules.PushIntervalSeconds), f.Cfg.Rules.PushJobName)
+	start := f.Prom.HandlerFunc()
+	start(ctx)
```
Comment (baerwang) [inline pixiu/pkg/filter/prometheus/metric.go]: should not ignore err

File: pixiu/pkg/filter/prometheus/metric.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package prometheus
+
+import (
+	nh "net/http"
```
Comment (baerwang) [inline pixiu/pkg/filter/prometheus/metric.go]: nh change stdHttp?

**Review by baerwang — COMMENTED** [review]
pls fix other question

**Review by baerwang — APPROVED** [review]

**Review by Lvnszn — APPROVED** [review]

**Review by mark4z — APPROVED** [review]

**Pull request merged**